x-pack/plugin/migrate/src/main/java/org/elasticsearch/xpack/migrate/action/ReindexDataStreamIndexTransportAction.java

```diff
@@ -0,0 +1,240 @@
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.cluster.block.ClusterBlockException;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.MetadataCreateIndexService;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.settings.IndexScopedSettings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.index.reindex.BulkByScrollResponse;
+import org.elasticsearch.index.reindex.ReindexAction;
+import org.elasticsearch.index.reindex.ReindexRequest;
+import org.elasticsearch.injection.guice.Inject;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class ReindexDataStreamIndexTransportAction extends HandledTransportAction<
+    ReindexDataStreamIndexAction.Request,
+    ReindexDataStreamIndexAction.Response> {
+
+    private static final Logger logger = LogManager.getLogger(ReindexDataStreamIndexTransportAction.class);
+
+    private static final Set<String> SETTINGS_TO_ADD_BACK = Set.of(IndexMetadata.SETTING_BLOCKS_WRITE, IndexMetadata.SETTING_READ_ONLY);
+
+    private static final IndicesOptions IGNORE_MISSING_OPTIONS = IndicesOptions.fromOptions(true, true, false, false);
+    private final ClusterService clusterService;
+    private final Client client;
+    private final IndexScopedSettings indexScopedSettings;
+
+    @Inject
+    public ReindexDataStreamIndexTransportAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ActionFilters actionFilters,
+        Client client,
+        IndexScopedSettings indexScopedSettings
+    ) {
+        super(
+            ReindexDataStreamIndexAction.NAME,
+            false,
+            transportService,
+            actionFilters,
+            ReindexDataStreamIndexAction.Request::new,
+            transportService.getThreadPool().executor(ThreadPool.Names.GENERIC)
+        );
+        this.clusterService = clusterService;
+        this.client = client;
+        this.indexScopedSettings = indexScopedSettings;
+    }
+
+    @Override
+    protected void doExecute(
+        Task task,
+        ReindexDataStreamIndexAction.Request request,
+        ActionListener<ReindexDataStreamIndexAction.Response> listener
+    ) {
+        var sourceIndexName = request.getSourceIndex();
+        var destIndexName = generateDestIndexName(sourceIndexName);
+        IndexMetadata sourceIndex = clusterService.state().getMetadata().index(sourceIndexName);
+        Settings settingsBefore = sourceIndex.getSettings();
+
+        if (sourceIndex.getCreationVersion().isLegacyIndexVersion() == false) {
```

Comment:
It sounds like `isLegacyIndexVersion` is not what we want to be using. Maybe use ReindexDataStreamAction::getOldIndexVersionPredicate instead?